apps/bfd-server/bfd-server-war/src/main/java/gov/cms/bfd/server/war/commons/QueryUtils.java

```diff
@@ -178,4 +221,49 @@ public static boolean isInRange(Instant lastUpdated, DateRangeParam range) {
     }
     return true;
   }
+
+  /**
+   * Query database to determine which claim types have data for the specified beneficiary.
+   *
+   * @param entityManager {@link EntityManager} used to query database.
+   * @param bene_id used to identify the Benficiary to chek claims for.
+   * @return {@link BitSet} denoting which claims have data.
+   */
+  public static BitSet hasClaimsData(EntityManager entityManager, long bene_id) {
+    /*
+     * execute a database function that returns a bitwise mask value that denotes that the given
+     * claim type will have data for the specified beneficiayrId. This represents fast and efficient
+     * way to ignore a requested claim type that ultimately has no data for our beneficiaryId.
+     *
+     * The database function returns bits as follows:
+     * CARRIER_CLAIMS      : bit 0
+     * INPATIENT CLAIMS    : bit 1
+     * OUTPATIENT CLAIMS   : bit 2
+     * SNF CLAIMS          : bit 3
+     * DME CLAIMS          : bit 4
+     * HHA CLAIMS          : bit 5
+     * HOSPICE CLAIMS      : bit 6
+     * PART D CLAIMS       : bit 7
+     *
+     * For more information on the database function, see: V111__SETUP_CLAIMS_AVAILABILITY_FUNCTION.SQL
+     * in the db migration directory.
+     */
+    List<Object> values =
+        entityManager
+            .createNativeQuery(CHECK_CLAIMS_FOR_DATA_SQL)
+            .setParameter("beneIdValue", bene_id)
+            .getResultList();
+
+    Integer maskVal = (Integer) (values != null && values.size() > 0 ? values.get(0) : 0);
+    BitSet rslt = new BitSet(maskVal);
+    rslt.set(CARRIER_HAS_DATA, (maskVal & V_CARRIER_HAS_DATA) == 1);
+    rslt.set(INPATIENT_HAS_DATA, (maskVal & V_INPATIENT_HAS_DATA) == 1);
+    rslt.set(OUTPATIENT_HAS_DATA, (maskVal & V_OUTPATIENT_HAS_DATA) == 1);
+    rslt.set(SNF_HAS_DATA, (maskVal & V_SNF_HAS_DATA) == 1);
+    rslt.set(DME_HAS_DATA, (maskVal & V_DME_HAS_DATA) == 1);
+    rslt.set(HHA_HAS_DATA, (maskVal & V_HHA_HAS_DATA) == 1);
+    rslt.set(HOSPICE_HAS_DATA, (maskVal & V_HOSPICE_HAS_DATA) == 1);
+    rslt.set(PART_D_HAS_DATA, (maskVal & V_PART_D_HAS_DATA) == 1);
+    return rslt;
```

Comment:
Since the integer bit mask is being processed here to create a `BitSet` why not create an `EnumSet` instead so that the providers could use a map or iterate over the enum values to decide what to do?  You could possibly even do an intersection of the `claimTypes` and result of this method.  `findByPatient` is begging to be a loop anyway and this could allow that.

Reply:
Excellent suggestion; there is follow-up work here to run the EOBs in parallel; will change the BitSet to EnumSet then. Also, cleaned up the mask stuff and added tests to assert mask value setting, etc. Thanks for review and suggestions...always appreciated.